### PR TITLE
CircleCI 2 config

### PR DESCRIPTION
**Pull request by nebril**
All basic test jobs are running in parallel, there could be some improvements here and there, but it works.

Not sure about deployment job, @adamreese @thomastaylor312 PTAL.

### CODE REVIEW

**Comment by thomastaylor312** [comment]
I'll take a look at this tonight. I'll give the deployment stuff a look too, but I am betting that @adamreese knows the most

File: .circleci/config.yml [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    steps:
+      - run:
+          name: update PATH
+          command: echo 'export PATH=$GOPATH/bin:$PATH' >> $BASH_ENV
+      - run:
+          name: remove old go files
+          command: rm -rf /usr/local/go && rm -rf "$GOPATH"
+      - run:
+          name: install go
+          command: wget "https://storage.googleapis.com/golang/go${GOVERSION}.linux-amd64.tar.gz" -O "${HOME}/go${GOVERSION}.tar.gz" && tar -C /usr/local -xzf "${HOME}/go${GOVERSION}.tar.gz"
```
Comment (adamreese) [inline .circleci/config.yml]: The installing go steps can be removed because it is running in the `golang:1.8` image

File: .circleci/config.yml [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    docker:
+      - image: golang:1.8
+    environment:
+      GOVERSION: "1.8.3"
+      GOPATH:  "/root/.go_workspace"
+      WORKDIR: "/root/.go_workspace/src/k8s.io/helm"
+      PROJECT_NAME: "kubernetes-helm"
+    steps:
+      - run:
+          name: update PATH
```
Comment (adamreese) [inline .circleci/config.yml]: Can all this path munging be removed now?

Comment (thomastaylor312) [inline .circleci/config.yml]: I had to manually add the Go bin in last time I did a Circle 2 file

File: .circleci/config.yml [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+          command: echo 'export PATH=$GOPATH/bin:$PATH' >> $BASH_ENV
+      - run:
```
Comment (thomastaylor312) [inline .circleci/config.yml]: Likewise here, the go installation steps can be pulled out

File: .circleci/config.yml [hunk 4]
```diff
@@ -0,0 +1,70 @@
+version: 2
+jobs:
+  build:
+    parallelism: 4
+    docker:
+      - image: golang:1.8
+    environment:
+      GOVERSION: "1.8.3"
+      GOPATH:  "/root/.go_workspace"
+      WORKDIR: "/root/.go_workspace/src/k8s.io/helm"
```
Comment (thomastaylor312) [inline .circleci/config.yml]: We won't need `WORKDIR` any more. We can now set the working directory with `working_directory`. See an example [here](https://github.com/thomastaylor312/helm-commandeer/blob/master/.circleci/config.yml#L4)

File: .circleci/config.yml [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+      - run:
+          name: remove old go files
+          command: rm -rf /usr/local/go && rm -rf "$GOPATH"
+      - run:
+          name: install go
+          command: wget "https://storage.googleapis.com/golang/go${GOVERSION}.linux-amd64.tar.gz" -O "${HOME}/go${GOVERSION}.tar.gz" && tar -C /usr/local -xzf "${HOME}/go${GOVERSION}.tar.gz"
+      - checkout
+      - run:
+          name: move repository to the canonical import path
+          command: mkdir -p "$(dirname ${WORKDIR})" && mv "${HOME}/project" "${WORKDIR}"
```
Comment (thomastaylor312) [inline .circleci/config.yml]: You won't need this with `working_directory` specified

File: .circleci/config.yml [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+      - run:
+          name: install dependencies
+          command: cd "${WORKDIR}" && make bootstrap
```
Comment (thomastaylor312) [inline .circleci/config.yml]: Won't need the `cd` any more with `working_directory`

File: .circleci/config.yml [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+          command: rm -rf /usr/local/go && rm -rf "$GOPATH"
+      - run:
+          name: install go
+          command: wget "https://storage.googleapis.com/golang/go${GOVERSION}.linux-amd64.tar.gz" -O "${HOME}/go${GOVERSION}.tar.gz" && tar -C /usr/local -xzf "${HOME}/go${GOVERSION}.tar.gz"
+      - checkout
+      - run:
+          name: move repository to the canonical import path
+          command: mkdir -p "$(dirname ${WORKDIR})" && mv "${HOME}/project" "${WORKDIR}"
+      - run:
+          name: install dependencies
```
Comment (thomastaylor312) [inline .circleci/config.yml]: I would cache the dependencies after they are installed and restore the cache in future build steps (unless we end up using workflows). See the docs [here](https://circleci.com/docs/2.0/configuration-reference/#save_cache)

File: .circleci/config.yml [hunk 8]
```diff
@@ -0,0 +1,70 @@
+version: 2
+jobs:
```
Comment (thomastaylor312) [inline .circleci/config.yml]: Obviously out of scope for this PR, but I wanted to capture somewhere that we should probably use `workflows` in the future to make some of these jobs (not steps) run at the same time

File: .circleci/config.yml [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+  build:
+    parallelism: 4
```
Comment (thomastaylor312) [inline .circleci/config.yml]: I am not sure if this is going to have the behavior we want. It is going to try and run each step of this job in parallel. See the docs [here](https://circleci.com/docs/2.0/parallelism-faster-jobs/) and [here](https://circleci.com/docs/2.0/configuration-reference/#build).

**Review by thomastaylor312 — CHANGES_REQUESTED** [review]
Some changes needed, but this looks pretty good. If you have any questions about my comments please let me know 🙂

File: .circleci/config.yml (hunk 9 above)
Comment (thomastaylor312) [inline .circleci/config.yml]: To clarify, I worry about it trying to run the tests and `make bootstrap` at the same time. Have you run into those issues?

File: .circleci/config.yml (hunk 1 above)
Comment (nebril) [inline .circleci/config.yml]: True.

File: .circleci/config.yml (hunk 9 above)
Comment (nebril) [inline .circleci/config.yml]: Each parallel build instance has it's own container, so it should be fine.

File: .circleci/config.yml (hunk 4 above)
Comment (nebril) [inline .circleci/config.yml]: True, will fix

File: .circleci/config.yml (hunk 3 above)
Comment (nebril) [inline .circleci/config.yml]: Will do

File: .circleci/config.yml (hunk 5 above)
Comment (nebril) [inline .circleci/config.yml]: Will fix

File: .circleci/config.yml (hunk 7 above)
Comment (nebril) [inline .circleci/config.yml]: Didn't find these docs, thanks!

File: .circleci/config.yml (hunk 6 above)
Comment (nebril) [inline .circleci/config.yml]: kk, thanks

**Comment by nebril** [comment]
@thomastaylor312 thanks for the feedback, fixes resulted in much smaller config file

File: .circleci/config.yml (hunk 8 above)
Comment (nebril) [inline .circleci/config.yml]: Makes sense, we can track it in an issue after this PR is merged

**Review by thomastaylor312 — APPROVED** [review]
This is looking great! Thanks @nebril! I'll wait for @adamreese to sign off too because he started the review

**Comment by adamreese** [comment]
Let merge after the 2.6.1 release

**Comment by adamreese** [comment]
@nebril Feel free to merge this, just ping me when you do

**Pull request merged**